inconsistent build instructions

at the moment shoggoth's `make` requires `sudo` so that `netlibc` can be installed in the system. this is an unexpected requirement as usually the UNIX build trifecta is `./configure && make && sudo make install` (you can remove the need for superuser on `install` entirely, if we let a binary prefix be selected e.g `make PREFIX=/home/user/.local`, but that's for the future).

this issue is definitely spun from https://github.com/shoggoth-systems/shoggoth/pull/11#issuecomment-1800606706, however giving my steps to reproduce my specific issues wouldn't work long-term. I believe having build instructions that closely follow the "usual UNIX build" would be beneficial to the project, as that would also help packaging the software on Linux distributions, or something like [reproducible builds](https://reproducible-builds.org/), but those would be long-term goals, beyond this issue.

for now, we can tackle creating a build path from an ubuntu or a void linux docker image (or anything else), as that would also let us run the test suite under Github Actions, especially to prevent regressions. I'm using void linux's docker images because it's what I run, which should also let you reproduce build failures I might have on my own system, Github Actions can come soon after we can pin this down.

1. create `Dockerfile.void` with following contents
```dockerfile
FROM ghcr.io/void-linux/void-glibc
# currently a hack because repo-default is very slow at this moment in time
RUN sed -i 's|https://repo-default.voidlinux.org|https://void.sakamoto.pl|g' /usr/share/xbps.d/00-repository-main.conf
RUN xbps-install -y -Suv
RUN xbps-install -y pkg-config python3 git gcc make which
ADD . /shoggoth
WORKDIR /shoggoth
RUN make
```
2. `docker build -t void-docker-shoggoth -f Dockerfile.void --progress=plain .`
3. currently fails with

```
#12 [8/8] RUN make
#12 0.315 which: no cargo in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin)
#12 0.721 netlibc header not found. Installing...
#12 0.722 make[2]: sudo: No such file or directory
#12 0.722 make[2]: *** [makefile:44: install] Error 127
#12 0.722 make[1]: *** [makefile:18: check-netlibc] Error 2
#12 0.723 make: *** [makefiles/build.mk:63: check-netlibc] Error 2
#12 ERROR: process "/bin/sh -c make" did not complete successfully: exit code: 2
------
```

moves the automatic script forward, so that's progress! next step would be deciding on what to do if `make` is without a terminal and without rust. my opinion is that it should fail and ask for rust installation from the system, rather than downloading precompiled binaries (either the .so or rustup would apply to this), relevant output:

```
#12 [8/8] RUN make
#12 0.257 which: no cargo in (/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin)
#12 0.266 Building redisjson...
#12 0.267 A dependency "redisjson" requires the Rust programming language to be installed
#12 0.268 choose one of the following options to proceed (recommend option 1) :
#12 0.269 1. Download a precompiled binary of "redisjson" without installing Rust, and continue the build
#12 0.270 2. Automatically install Rust and build "redisjson" from source using ./lib/redisjson/
#12 0.271 3. Cancel the build and exit
#12 0.272 Invalid choice:
#12 0.272 make: *** [makefiles/build.mk:114: target/redisjson.so] Error 11
#12 ERROR: process "/bin/sh -c make" did not complete successfully: exit code: 2
------
```

Now the dockerfile is as follows (added openssl headers and libuuid):
```dockerfile
FROM ghcr.io/void-linux/void-glibc
RUN sed -i 's|https://repo-default.voidlinux.org|https://void.sakamoto.pl|g' /usr/share/xbps.d/00-repository-main.conf
RUN xbps-install -y -Suv
RUN xbps-install -y pkg-config python3 git gcc make
RUN xbps-install -y which
RUN xbps-install -y openssl openssl-devel
RUN xbps-install -y libuuid libuuid-devel
ADD . /shoggoth
WORKDIR /shoggoth
RUN make
```

And this is the build output
```
#14 [10/10] RUN make
#14 0.439 Building Shoggoth (flat)...
#14 4.678 Building libshogdb...
#14 4.751 Linking Shoggoth (flat)...
#14 4.884 Packaging v0.1.5 for Linux x86_64
#14 4.904 Building Shoggoth (debug)...
#14 7.387 Creating Shoggoth static lib (debug)...
#14 8.348 Building explorer...
#14 8.533 Generating explorer ...
#14 8.547 Building Shoggoth (debug)...
#14 10.67 Creating Shoggoth static lib (debug)...
#14 11.92 Building docs...
#14 12.92 Generating docs ...
#14 12.92 Building shogdb...
#14 13.79 /usr/sbin/ld: cannot find libasan_preinit.o: No such file or directory
#14 13.79 /usr/sbin/ld: cannot find -lasan: No such file or directory
#14 13.79 /usr/sbin/ld: cannot find -lasan: No such file or directory
#14 13.79 /usr/sbin/ld: cannot find -lubsan: No such file or directory
#14 13.81 /usr/sbin/ld: cannot find -lubsan: No such file or directory
#14 13.81 collect2: error: ld returned 1 exit status
#14 13.81 make[3]: *** [makefile:43: target/shogdb] Error 1
#14 13.81 make[2]: *** [makefiles/build.mk:117: target/shogdb] Error 2
#14 13.82 make[1]: *** [makefiles/package.mk:102: create-package] Error 2
#14 13.82 make: *** [makefiles/package.mk:70: package-release] Error 2
#14 ERROR: process "/bin/sh -c make" did not complete successfully: exit code: 2
```

--- a/README.txt
+++ b/README.txt
@@ -16,6 +16,24 @@ To learn more about Shoggoth, read the documentation at https://shoggoth.network
 
 You can install Shoggoth by following the instructions at https://shoggoth.network/explorer/docs#installation
 
+BUILD
+================================
+
+The following dependencies are required to build Shoggoth
+- gcc
+- GNU make
+- pkg-config
+- which
+- openssl (and its development headers)
+- libuuid (and its development headers)
+- libsanitizer (and its development headers)
+- GNU find
+- zip
+
+If on Void Linux, the following command should install the dependencies:
+`xbps-install pkg-config gcc make which openssl openssl-devel libuuid libuuid-devel libsanitizer-devel findutils zip`
+
+Then, run `make` on the shoggoth directory. You will have Shoggoth at `./target/shog-flat`.
 
 LICENSE
 =================================

--- a/README.txt
+++ b/README.txt
@@ -16,6 +16,8 @@ To learn more about Shoggoth, read the documentation at https://shoggoth.network
 
 You can install Shoggoth by following the instructions at https://shoggoth.network/explorer/docs#installation
 
+
+
 BUILD
 ================================
 
@@ -39,11 +41,13 @@ $ make
 
 This will build Shoggoth with supporting runtime files into a zip file ./target/shoggoth-release-<VERSION>-<OS>-<ARCH>.zip for example ./target/shoggoth-release-v0.1.5-Linux-x86_64.zip
 
-After building, to install Shoggoth for the current user run:
+After building, to install Shoggoth for the current user, run:
 
 $ make install
 
 NOTE: do not use sudo. Shoggoth will not be installed globally, so sudo is not required.
+
+
 
 LICENSE
 =================================
@@ -52,6 +56,7 @@ Shoggoth is licensed under the MIT license. Read the LICENSE file for more infor
 
 Shoggoth uses a few dependencies which have their own licenses. The dependencies in the ./lib/ directory are independent of the MIT license used for Shoggoth.
 The source code for each dependency includes a LICENSE file indicating the license that covers it.
+
 
 
 LINKS

--- a/README.txt
+++ b/README.txt
@@ -20,20 +20,30 @@ BUILD
 ================================
 
 The following dependencies are required to build Shoggoth
-- gcc
+- gcc or clang
 - GNU make
 - pkg-config
-- which
 - openssl (and its development headers)
 - libuuid (and its development headers)
 - libsanitizer (and its development headers)
 - GNU find
 - zip
 
-If on Void Linux, the following command should install the dependencies:
-`xbps-install pkg-config gcc make which openssl openssl-devel libuuid libuuid-devel libsanitizer-devel findutils zip`
+If building on a UNIX system like Linux or macOS, run the configure script to automatically install the above dependencies:
 
-Then, run `make` on the shoggoth directory. You will have Shoggoth at `./target/shog-flat`.
+$ ./configure
+
+Then, run `make` on the shoggoth directory:
+
+$ make
+
+This will build Shoggoth with supporting runtime files into a zip file ./target/shoggoth-release-<VERSION>-<OS>-<ARCH>.zip for example ./target/shoggoth-release-v0.1.5-Linux-x86_64.zip
+
+After building, to install Shoggoth for the current user run:
+
+$ make install
+
+NOTE: do not use sudo. Shoggoth will not be installed globally, so sudo is not required.
 
 LICENSE
 =================================

